### PR TITLE
Fix two potential OOM issues in GPU aggregate.

### PR DESCRIPTION
**Pull request by firestarman**
close https://github.com/NVIDIA/spark-rapids/issues/11903

The first one is by taking the nested literals into account when calculating the output size for pre-split. See the linked issue above for more details.

The second one is by using the correct size for buffer size comparison when collecting the next bundle of batches in aggregate. The size return from the `batchesByBucket.last.size()` is not the actual buffer size in bytes,
but the element number of an array. It can not be used for the buffer size comparison.

I verified this PR locally by the toy query and it works well.

### CODE REVIEW

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuAggregateExec.scala [hunk 1]
```diff
@@ -1095,16 +1095,13 @@ class GpuMergeAggregateIterator(
...
-                  // for some test cases targetMergeBatchSize is too small to fit any bucket,
-                  // in this case we put the first bucket into toAggregateBuckets anyway
-                  // refer to https://github.com/NVIDIA/spark-rapids/issues/11790 for examples
-                  toAggregateBuckets.isEmpty ||
-                    batchesByBucket.last.size() + currentSize <= targetMergeBatchSize)) {
-                val bucket = batchesByBucket.remove(batchesByBucket.size - 1)
-                currentSize += bucket.map(_.sizeInBytes).sum
-                toAggregateBuckets += bucket
+              var keepGoing = true
+              while (batchesByBucket.nonEmpty && keepGoing) {
```
Comment (binmahone) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuAggregateExec.scala:1099]: may need use a separate PR for this as it is irrelevant to the description in https://github.com/NVIDIA/spark-rapids/issues/11903

Comment (firestarman) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuAggregateExec.scala:1099]: Thx for review. @binmahone Could you help file an issue for this? Then I will follow your suggestion.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala [hunk 2]
```diff
@@ -308,12 +311,97 @@ object PreProjectSplitIterator {
...
+
+  private def calcMemorySizeForLiteral(litVal: Any, litType: DataType, numRows: Int): Long = {
+    // One literal size = value size + additional buffers size (offset and/or validity)
+    val litSize = calcLitValueSize(litVal, litType) + {
+      val pickRowNum: Int => Int = rowNum => if (litVal == null) 0 else rowNum
+      litType match {
+        case ArrayType(elemType, hasNullElem) =>
+          val numElems = pickRowNum(litVal.asInstanceOf[ArrayData].numElements())
+          // A GPU array literal requires only one column as the child
+          estimateLitAdditionSize(hasNullElem, hasOffset(elemType), numElems)
```
Comment (binmahone) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala:334]: is there a risk of over estimation here?

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuAggregateExec.scala (hunk 1 above)
Comment (binmahone) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuAggregateExec.scala:1099]: you can also consider modifying  the description in 11903 :-)

Comment (firestarman) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuAggregateExec.scala:1099]: Updated

**Comment by firestarman** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala [hunk 3]
```diff
@@ -308,12 +311,97 @@ object PreProjectSplitIterator {
...
+      litTp.defaultSize
+    } else {
+      0L
+    }
+  } else {
+    litTp match {
+      case StringType => lit.asInstanceOf[UTF8String].numBytes()
+      case BinaryType => lit.asInstanceOf[Array[Byte]].length
+      case ArrayType(elemType, _) =>
+        lit.asInstanceOf[ArrayData].array.map(calcLitValueSize(_, elemType)).sum
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala:381]: This appears to only work on UnsafeArrayData. If I test with the following it fails. (I opened up the API to be public so I could test things manually)

```
scala> PreProjectSplitIterator.calcMemorySizeForLiteral(ArrayData.toArrayData(Array(1L)), ArrayType(LongType), 100)
java.lang.UnsupportedOperationException: Not supported on UnsafeArrayData.
  at org.apache.spark.sql.catalyst.expressions.UnsafeArrayData.array(UnsafeArrayData.java:103)
  at com.nvidia.spark.rapids.PreProjectSplitIterator$.calcLitValueSize(basicPhysicalOperators.scala:381)
  at com.nvidia.spark.rapids.PreProjectSplitIterator$.calcMemorySizeForLiteral(basicPhysicalOperators.scala:328)
  ... 51 elided
```

**Review by revans2 — CHANGES_REQUESTED** [review]
This needs a lot of testing in different cases to be sure that it works properly. Especially for nested types.

**Comment by firestarman** [comment]
Thx for review, I am doing a small refactor for the meta size calculation part... Will update it once done.

**Comment by firestarman** [comment]
Update for early review, I will run some tests next.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala [hunk 4]
```diff
@@ -308,12 +311,97 @@ object PreProjectSplitIterator {
...
+          // A GPU array literal requires only one column as the child
+          estimateLitAdditionSize(hasNullElem, hasOffset(elemType), numElems)
+        case StructType(fields) =>
+          val childrenNumRows = pickRowNum(1)
+          // A GPU struct literal requires "fields.size" columns as the children.
+          fields.map(f =>
+            estimateLitAdditionSize(f.nullable, hasOffset(f.dataType), childrenNumRows)
+          ).sum
+        case MapType(keyType, valType, hasNullValue) =>
+          val mapRowsNum = pickRowNum(litVal.asInstanceOf[MapData].numElements())
```
[the remark this answers is not in the document]
Comment (firestarman) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala]: Updated by the refactor.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala [hunk 5]
```diff
@@ -308,12 +311,97 @@ object PreProjectSplitIterator {
...
     }.sum
   }
+
+  private def calcMemorySizeForLiteral(litVal: Any, litType: DataType, numRows: Int): Long = {
+    // One literal size = value size + additional buffers size (offset and/or validity)
+    val litSize = calcLitValueSize(litVal, litType) + {
+      val pickRowNum: Int => Int = rowNum => if (litVal == null) 0 else rowNum
+      litType match {
+        case ArrayType(elemType, hasNullElem) =>
+          val numElems = pickRowNum(litVal.asInstanceOf[ArrayData].numElements())
```
[the remark this answers is not in the document]
Comment (firestarman) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala]: Updated by the refactor.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala (hunk 2 above)
Comment (firestarman) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala]: I have done the refactor to calculate the size more exactly.
I follow the type definitions from  https://github.com/rapidsai/cudf/blob/a0487be669326175982c8bfcdab4d61184c88e27/cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md#list-columns

**Comment by firestarman** [comment]
> Looks good to me. I still would like to see some tests (at least some simple unit tests) to validate that it is doing the right thing and that we are covering corner cases like nulls, and arrays of arrays.

Sure, working on the tests now...

**Comment by firestarman** [comment]
Added some tests, also found some bugs and fixed them.

**Comment by firestarman** [comment]
build

**Review by revans2 — APPROVED** [review]
